### PR TITLE
[fix] [broker] reader stuck when read from compacted topic with read compact mode disable

### PR DESCRIPTION
**Pull request by thetumbled**
### Motivation
`reader.hasMessageAvailable()` rely on `getLastMessageId` to determine whether there are any messages available to read.
`org.apache.pulsar.broker.service.ServerCnx#handleGetLastMessageId` will read the last message from compacted topic in some cases without check if the consumer is in read compact mode. 
If the consumer/reader disable the read compact mode, and the cluster enable the compaction feature, messages will be compacted into the compacted topic, but the dispactcher will not dispatch messages from compacted topic to consumer. If the messages in the original topic is expired and the consumer has nothing to be read, while the messages in the compacted topic retain alive. 
**`reader.hasMessageAvailable()` will return true as there are messages in the compacted topic, but consumer/reader will not get any messages because there are nothing in the original topic.**

### Modifications
`reader.hasMessageAvailable()`  should check if the consumer is in read compact mode to determine whether get the last message id from the compacted topic.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.
This change added tests and can be verified as follows:

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/thetumbled/pulsar/pull/38

### CODE REVIEW

**Review by coderzc — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 1]
```diff
@@ -2213,8 +2216,23 @@ private void getLargestBatchIndexWhenPossible(
...
+                    // if readCompacted is true, we need to read the last entry from compacted topic
+                    handleLastMessageIdFromCompactionService(persistentTopic, requestId, partitionIndex,
+                            markDeletePosition);
+                } else {
+                    // if readCompacted is false, we need to return MessageId.earliest
+                    writeAndFlush(Commands.newGetLastMessageIdResponse(requestId, -1, -1, partitionIndex, -1,
+                            markDeletePosition != null ? markDeletePosition.getLedgerId() : -1,
+                            markDeletePosition != null ? markDeletePosition.getEntryId() : -1));
+                }
+                return;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:2231]: nit:

```diff --git a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java
index ab4758d5be..3ab25eb098 100644
--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java
@@ -2218,7 +2218,7 @@ public class ServerCnx extends PulsarHandler implements TransportCnx {
 
             if (lastPosition.getEntryId() == -1 || !ml.ledgerExists(lastPosition.getLedgerId())) {
                 // there is no entry in the original topic
-                if (readCompacted) {
+                if (compactionHorizon != null) {
                     // if readCompacted is true, we need to read the last entry from compacted topic
                     handleLastMessageIdFromCompactionService(persistentTopic, requestId, partitionIndex,
                             markDeletePosition);
@@ -2231,8 +2231,7 @@ public class ServerCnx extends PulsarHandler implements TransportCnx {
                 return;
             }
 
-            if (readCompacted && compactionHorizon != null
-                    && lastPosition.compareTo((PositionImpl) compactionHorizon) <= 0) {
+            if (compactionHorizon != null && lastPosition.compareTo((PositionImpl) compactionHorizon) <= 0) {
                 handleLastMessageIdFromCompactionService(persistentTopic, requestId, partitionIndex,
                         markDeletePosition);
                 return;
```

**Review by codelipenghui — APPROVED** [review]
Nice catch!

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java (hunk 1 above)
Comment (thetumbled) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:2231]: fixed.

**Review by Technoboy- — APPROVED** [review]

**Comment by heesung-sn** [comment]
@thetumbled can you help to cherry-pick this PR to branch-3.0? I see conflicts.

**Comment by thetumbled** [comment]
> @thetumbled can you help to cherry-pick this PR to branch-3.0? I see conflicts.

We have better wait https://github.com/apache/pulsar/pull/22130 to be merged, or two pr will conflict. I will help to cherry pick this pr.

**Comment by thetumbled** [comment]
> @thetumbled can you help to cherry-pick this PR to branch-3.0? I see conflicts.
cherry pick in this pr: https://github.com/apache/pulsar/pull/22199

**Pull request merged**